File: C/hardware/stream_to_disk_example/stream_to_disk_example.c
//////////////////////////////////////////////////////////////////
//
// stream_to_disk_example.c - C file
//
// This example demonstrates the double buffering capabilities of the C API.
// It reads continuously from analog input channels 0 and 1 and encoder
// channels 0 and 1, at a sampling rate of 1kHz. The data is written to
// a file as it is collected.
//
// Data collection is started using the hil_task_start function.
// The data is then read from the C API's internal buffer in one second
// intervals (1000 samples), using the hil_task_read function.
//
// Stop the example by pressing Ctrl-C. 
//
// This example demonstrates the use of the following functions:
//    hil_open
//    hil_task_create_reader
//    hil_task_start
//    hil_task_read
//    hil_task_stop
//    hil_task_delete
//    hil_close
//
// Copyright (C) 2008 Quanser Inc.
//////////////////////////////////////////////////////////////////
    
#include "stream_to_disk_example.h"

static int stop = 0;

void signal_handler(int signal)
{
    stop = 1;
}

int main(int argc, char * argv[])
{
    static const char board_type[]       = "q2_usb";
    static const char board_identifier[] = "0";
    static char       message[512];

    qsigaction_t action;
    t_card board;
    t_int  result;

    /* Catch Ctrl+C to shut down application cleanly */
    action.sa_handler = signal_handler;
    action.sa_flags   = 0;
    qsigemptyset(&action.sa_mask);

    qsigaction(SIGINT, &action, NULL);

    result = hil_open(board_type, board_identifier, &board);
    if (result == 0)
    {
        const t_uint32  samples              = -1; /* read continuously */
        const t_uint32  analog_channels[]    = { 0, 1 };
        const t_uint32  encoder_channels[]   = { 0, 1 };
        const t_double  frequency            = 1000;

        #define NUM_ANALOG_CHANNELS     ARRAY_LENGTH(analog_channels)
        #define NUM_ENCODER_CHANNELS    ARRAY_LENGTH(encoder_channels)
        #define SAMPLES_TO_READ         1000    /* one second's worth */

        const t_uint32  samples_in_buffer    = (t_uint32)(2 * SAMPLES_TO_READ); /* double buffer */
        const t_double  period               = 1.0 / frequency;

        static t_int32  counts[SAMPLES_TO_READ][NUM_ENCODER_CHANNELS];
        static t_double voltages[SAMPLES_TO_READ][NUM_ANALOG_CHANNELS];
        static const char default_filename[] = "data.txt";

        t_int  samples_read;
        t_task task;
        char filename[_MAX_PATH];
        FILE * file_handle;

        printf("This example reads the first two analog input channels and encoder channels\n");
        printf("at %g Hz, continuously, writing the data to a file.\n", frequency);
        
        printf("Press CTRL-C to stop reading.\n\n");
        printf("Enter the name of the file to which to write the data [%s]:\n", default_filename);
        fgets(filename, ARRAY_LENGTH(filename), stdin); /* use fgets to avoid deprecation warnings that occur on some systems */
        filename[string_length(filename, sizeof(filename)) - 1] = '\0';
        if (filename[0] == '\0')
            string_copy(filename, sizeof(filename), default_filename);
        
        if (stdfile_open(filename, "wt", &file_handle) == 0)
        {
            t_double time = 0;

            printf("\nWriting to the file. Each dot represents %d data points.\n", SAMPLES_TO_READ);

            result = hil_task_create_reader(board, samples_in_buffer, analog_channels, NUM_ANALOG_CHANNELS,
                encoder_channels, NUM_ENCODER_CHANNELS, NULL, 0, NULL, 0, &task);
            if (result == 0)
            {
                result = hil_task_start(task, HARDWARE_CLOCK_0, frequency, samples);
                if (result == 0)
                {
                    samples_read = hil_task_read(task, SAMPLES_TO_READ, &voltages[0][0], &counts[0][0], NULL, NULL);
                    while (samples_read > 0 && stop == 0)
                    {
                        t_uint channel;
                        t_int index;

                        printf(".");
                        fflush(stdout);

                        for (index = 0; index < samples_read; index++)
                        {
                            fprintf(file_handle, "t: %8.4f  ", time);
                            time += period;

                            for (channel = 0; channel < NUM_ANALOG_CHANNELS; channel++)
                                fprintf(file_handle, "ADC #%d: %5.3f    ", analog_channels[channel], voltages[index][channel]);
                            for (channel = 0; channel < NUM_ENCODER_CHANNELS; channel++)
                                fprintf(file_handle, "ENC #%d: %5d    ", encoder_channels[channel], counts[index][channel]);
                            fprintf(file_handle, "\n");
                        }

                        samples_read = hil_task_read(task, SAMPLES_TO_READ, &voltages[0][0], &counts[0][0], NULL, NULL);
                    }

                    hil_task_stop(task);

                    if (samples_read < 0)
                    {
                        msg_get_error_message(NULL, samples_read, message, ARRAY_LENGTH(message));
                        printf("Unable to read channels. %s Error %d.\n", message, -samples_read);
                    }
                    else
                    {
                        printf("\nRead operation has been stopped. Press Enter to continue.\n");
                        getchar(); /* absorb Ctrl+C */
                    }
                }
                else
                {
                    msg_get_error_message(NULL, result, message, ARRAY_LENGTH(message));
                    printf("Unable to start task. %s Error %d.\n", message, -result);
                }

                hil_task_delete(task);
            }
            else
            {
                msg_get_error_message(NULL, result, message, ARRAY_LENGTH(message));
                printf("Unable to create task. %s Error %d.\n", message, -result);
            }

            stdfile_close(file_handle);
        }
        else
            printf("Unable to open file \"%s\".\n", filename);

        hil_close(board);
    }
    else
    {
        msg_get_error_message(NULL, result, message, ARRAY_LENGTH(message));
        printf("Unable to open board. %s Error %d.\n", message, -result);
    }

    return 0;
}

File: python/hardware/stream_to_disk_example/stream_to_disk_example.py
#################################################################
#
# stream_to_disk_example.py - Python file
#
# This example demonstrates the double buffering capabilities of the Python API.
# It reads continuously from analog input channels 0 and 1 and encoder
# channels 0 and 1, at a sampling rate of 1 kHz. The data is written to
# a file as it is collected.
#
# Data collection is started using the HIL.task_start function.
# The data is then read from the Python API's internal buffer in one second
# intervals (1000 samples), using the HIL.task_read function.
#
# Stop the example by pressing Ctrl-C. 
#
# This example demonstrates the use of the following functions:
#    HIL.open
#    HIL.task_create_reader
#    HIL.task_start
#    HIL.task_read
#    HIL.task_stop
#    HIL.task_delete
#    HIL.close
#
# Copyright (C) 2023 Quanser Inc.
#################################################################

from quanser.hardware import HIL, HILError, Clock
import signal
import sys
import array as arr

stop = False

def signal_handler(signum, frame):
    global stop
    stop = True

# Register a Ctrl+C handler
signal.signal(signal.SIGINT, signal_handler)

board_type = "q2_usb"
board_identifier = "0"

try:
    card = HIL()
    card.open(board_type, board_identifier)

    try:
        samples           = HIL.INFINITE # read continuously
        samples_to_read   = 1000
        frequency         = 1000
        samples_in_buffer = 2 * samples_to_read # double buffer
        period            = 1 / frequency
        default_filename  = "data.txt"

        analog_channels   = arr.array("i", [0])
        encoder_channels  = arr.array("i", [0, 1])

        num_analog_channels  = len(analog_channels)
        num_encoder_channels = len(encoder_channels)

        voltages = arr.array("d", [0.0] * samples_to_read * num_analog_channels)
        counts   = arr.array("i", [0] * samples_to_read * num_encoder_channels)

        print("This example reads the first two analog input channels and encoder channels")
        print("at %g Hz, continuously, writing the data to a file." % frequency)
        
        print("Press CTRL-C to stop reading.\n");
        filename = input("Enter the name of the file to which to write the data [" + default_filename + "]:")
        if len(filename) == 0:
            filename = default_filename

        file_handle = open(filename, "wt")

        time = 0

        print("\nWriting to the file. Each dot represents %d data points." % samples_to_read)
        
        task = card.task_create_reader(samples_in_buffer, analog_channels, num_analog_channels,
            encoder_channels, num_encoder_channels, None, 0, None, 0)

        card.task_start(task, Clock.HARDWARE_CLOCK_0, frequency, samples)

        samples_read = card.task_read(task, samples_to_read, voltages, counts, None, None)
        while samples_read > 0 and stop == False:
            print(".", end="", flush=True)

            for index in range(samples_read):
                print("t: %8.4f" % time, end="  ", file=file_handle)
                time += period

                for channel in range(num_analog_channels):
                    print("ADC #%d: %5.3f" % (analog_channels[channel], voltages[index * num_analog_channels + channel]), end="  ", file=file_handle)
                for channel in range(num_encoder_channels):
                    print("ENC #%d: %5d" % (encoder_channels[channel], counts[index * num_encoder_channels + channel]), end="  ", file=file_handle)
                print(flush=True, file=file_handle)

            samples_read = card.task_read(task, samples_to_read, voltages, counts, None, None)

        card.task_stop(task)
        card.task_delete(task)

    except HILError as ex:
        print("\nUnable to run task. %s" % ex.get_error_message())

    finally:
        card.close()

except HILError as ex:
    print("Unable to open board. %s" % ex.get_error_message())

input("\nPress Enter to continue.")
exit(0)
